daemon/uevent.c

```diff
@@ -298,6 +311,111 @@ uevent_get_usb_product(struct uevent *uevent)
 	return id_product;
 }
 
+/**
+ * This function forks a new child in the target netns (and userns) of netns_pid
+ * in which the uevents should be injected. In the child the UEVENT netlink socket
+ * is connected and a new message containing the raw uevent will be created and
+ * sent to that socket.
+ */
+static int
+uevent_inject_into_netns(char *uevent, size_t size, pid_t netns_pid, bool join_userns)
+{
+	int status;
+	pid_t pid = fork();
+
+	if (pid == -1) {
+		ERROR_ERRNO("Could not fork for switching to netns of %d", netns_pid);
+		return -1;
+	} else if (pid == 0) {
+		if (join_userns) {
+			char *usrns = mem_printf("/proc/%d/ns/user", netns_pid);
+			int usrns_fd = open(usrns, O_RDONLY);
+			mem_free(usrns);
+			if (usrns_fd == -1)
+				FATAL_ERRNO("Could not open userns file %s!", usrns);
+			if (setns(usrns_fd, CLONE_NEWUSER) == -1)
+				FATAL_ERRNO("Could not join uesr namespace of pid %d!", netns_pid);
+			if (setuid(0) < 0)
+				FATAL_ERRNO("Could setuid to root in user namespace of pid %d!",
+					    netns_pid);
+			if (setgid(0) < 0)
+				FATAL_ERRNO("Could setgid to root in user namespace of pid %d!",
+					    netns_pid);
+			if (setgroups(0, NULL) < 0)
+				FATAL_ERRNO("Could setgroups to root in user namespace of pid %d!",
+					    netns_pid);
+		}
+		char *netns = mem_printf("/proc/%d/ns/net", netns_pid);
+		int netns_fd = open(netns, O_RDONLY);
+		mem_free(netns);
+		if (netns_fd == -1)
+			FATAL_ERRNO("Could not open netns file %s!", netns);
+		if (setns(netns_fd, CLONE_NEWNET) == -1)
+			FATAL_ERRNO("Could not join network namespace of pid %d!", netns_pid);
+		nl_sock_t *target = nl_sock_uevent_new(0);
+		if (NULL == target)
+			FATAL("Could not connect to nl socket!");
+		nl_msg_t *nl_msg = nl_msg_new();
+		if (NULL == nl_msg)
+			FATAL_ERRNO("Could not allocate nl_msg!");
+		if (nl_msg_set_type(nl_msg, UEVENT_SEND) < 0)
+			FATAL("Could not set type UEVENT_SEND of nl_msg!");
+		if (nl_msg_set_flags(nl_msg, NLM_F_ACK | NLM_F_REQUEST))
+			FATAL("Could not set flages for acked request of nl_msg!");
+		if (nl_msg_set_buf_unaligned(nl_msg, uevent, size) < 0)
+			FATAL_ERRNO("Could not add uevent to nl_msg!");
+		if (nl_msg_send_kernel(target, nl_msg) < 0)
+			FATAL_ERRNO("Could not inject uevent!");
+		if (nl_msg_receive_and_check_kernel(target))
```

Comment:
nl_msg_receive_and_check_kernel() returns 0 also if error in nl was non-zero

Reply:
done